### ISSUE TITLE
Type assertions & type predicates can have incorrect narrowing logic

### ISSUE THREAD

**Issue by Masstronaut** [microsoft/TypeScript#57436]
### 🔎 Search Terms

Type assertion, type predicates, type narrowing

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about assertion functions, type predicates, and narrowing. I failed to find anything covering type predicates and assertion functions.

### ⏯ Playground Link

[see on TS playground](
https://www.typescriptlang.org/play?jsx=0#code/C4TwDgpgBAggdgSwLYEMA2UC8UDeU0QDmAzgFxRwCuSARhAE5QC+A3AFCiRQDCKwWsRKgwAyXPiJkoAFgA0UJBAD2Ad3IAKAJRYAfFABuShABNm7TtABCCeqezxk6KGJwES5AEzyAxgAsbYBramHqGJmZsbAD0UVAAKv7EUABmlHDewAhKcFDGShDEcADk-L4o+tDElMnJCN4IEHD8aEqEdVDAShQo9PSqKX1IUAAG1rZQAD48fMMdXaM2xsNsqemZ2VAIxGPG6pDA5DuT08Ca5PubSUc4bACQ9BDAlPQ5wPSUEOxM0bEJWylpDJZHJ5ArFUrlSrVWr1RrNVrtTrdXr9ZKDEZHKa8YCzJELWzLVZAjZbHYwYjEBjrOB7R6HRbHbFnKAoClUpIXf7XNjfFaA6mbOBVGp1BpNcmU+jU4i0g5QTEnTQ3KCbZLqUmLWWaJVsFUq-YAOj8AS07BV3xVGtsEqpwK1ZqghuN9DAWh5kRi8V80AsxG89AQYH43iUSDACAIjG8BB6aBAUDKSWA3oUEBQQrmCfTxgIyL6KgQcEIUDowBUEEaHW9lI64AKpD5a2BBnQJgAcj184XCLL6eMsXwdSrPcnKihFPgEd4WUlhlalpcoABrOCqV5dOiC4UwsX8JFwTv9Ue1rhKZIjfbLS1qt4fId6z0AAWAxAAtBAAB6QDLvlH0TwAMwAQAnLqepOv4LqmmBvLXrKBpuEkmDIVAHj3sOsQAOoIMmjp9JAjAHiiBZFpObTePIx4hmGEYMMuq4qEmKYWFA3jpiW0BEfmECmHiOxXuBjxGpBrqaA63xMEAA)

### 💻 Code

```ts
type Animal = { legs: number };
type Cat = Animal & { legs: 4, meow: () => void };
type Bird = Animal & {legs: 2, chirp: () => void };

// This function doesn't have sufficient logic to narrow from `Bird | Cat` to `Bird`
function isBird(pet: Bird | Cat): pet is Bird {
	return true;
}
// This function doesn't have sufficient logic to narrow from `Bird | Cat` to `Bird`
function isBirdAssertion(pet: Bird | Cat): asserts pet is Bird {
}

function insufficientAssertions(pet: Bird | Cat){
  if(isBird(pet)){
    pet.chirp();
  }
  isBirdAssertion(pet);
  pet.chirp()
}

// The typescript compiler clearly has the means to handle narrowing between these types:
function validNarrowing(pet: Bird | Cat){
  // the same logic as `isBird` is known to be insufficient to narrow the type of `pet`
  if(true){
    // @ts-expect-error: 2339
    pet.chirp();
  }

  if(pet.legs === 2){
    // With proper narrowing logic, the compiler knows the type can be narrowed to `Bird`
    pet.chirp();
  }
}
```

### 🙁 Actual behavior

An assertion function or type predicate which contains insufficient logic to narrow the input type generates no errors and can be used to narrow a type in the calling code, despite not having certainty the type is narrower.

### 🙂 Expected behavior

an assertion function or type predicate which doesn't sufficiently narrow the input type before the return or end of function should result in an error.

**Comment by fatcerberus** [comment]
The primary reason to write a UDTG is to do type checks that the compiler isn't capable of doing itself, so it's very much intentional that the implementation isn't checked.  Otherwise you'd get spurious errors in cases where the logic *is* sufficient for narrowing but the compiler isn't able to prove it.

**Comment by Masstronaut** [comment]
> The primary reason to write a UDTG is to do type checks that the compiler isn't capable of doing itself, so it's very much intentional that the implementation isn't checked.  Otherwise you'd get spurious errors in cases where the logic *is* sufficient for narrowing but the compiler isn't able to prove it.

Could you share an example? The TS handbook doesn't seem to imply that, and in fact gives an example similar to mine that the compiler definitely was capable of doing itself. [See "using type predicates"](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) for example.

Where is it more useful for type predicates to not also have the compiler perform the checks it can?

**Comment by fatcerberus** [comment]
> I have yet to encounter a scenario where it is more useful for type predicates to not also have the compiler perform the checks it can.

You’re basically asking TS to spit out an error whenever it thinks that you “haven’t done enough” to narrow the value according to the function signature, via the checks it’s able to do itself.  Lots of non-trivial type guards would be broken by such an error; see https://github.com/microsoft/TypeScript/issues/29980#issuecomment-467945410.

tl;dr: The proposed behavior would error on any type predicate where the same check wouldn’t already work to narrow the value inline.  In other words, almost all of them.

**Comment by tadhgmister** [comment]
An example that may be useful for the docs is when the property to narrow on 

```ts
interface SuccessStatus {
    code: 0;
    message: string;
}
interface FailureStatus {
    /* in failure case this will be an integer other than 0 */
    code: number;
    message: string;
}
type ProcessStatus = SuccessStatus | FailureStatus;

function isFailure(response: ProcessStatus): response is FailureStatus {
    return response.code !== 0;
}
```

Types like "any number other than this specific one" can't be explicitly expressed in typescript and therefore can't be automatically narrowed.

**Comment by tadhgmister** [comment]
example that may be useful for docs:

```ts
interface SuccessStatus {
    code: 0;
    message: string;
}
interface FailureStatus {
    /* in failure case code will be anything other than 0 */
    code: number;
    reason: string;
}
type ProcessStatus = SuccessStatus | FailureStatus;

function isFailure(response: ProcessStatus): response is SuccessStatus {
    return response.code === 0;
}

declare const a: ProcessStatus;

if (a.code === 0){
    // this doesn't work since typescript can't determine that failure case can't be 0
    console.log(a.message); 
}
if (isFailure(a)){
    console.log(a.message); // works as intended.
}
```

**Comment by Masstronaut** [comment]
> example that may be useful for docs:
> 
> ```ts
> interface SuccessStatus {
>     code: 0;
>     message: string;
> }
> interface FailureStatus {
>     /* in failure case code will be anything other than 0 */
>     code: number;
>     reason: string;
> }
> type ProcessStatus = SuccessStatus | FailureStatus;
> 
> function isSuccess(response: ProcessStatus): response is SuccessStatus {
>     return response.code === 0;
> }
> 
> declare const a: ProcessStatus;
> 
> if (a.code === 0){
>     // this doesn't work since typescript can't determine that failure case can't be 0
>     console.log(a.message); 
> }
> if (isSuccess(a)){
>     console.log(a.message); // works as intended.
> }
> ```

@tadhgmister In the example you shared the type could be narrowed safely:

```ts
interface SuccessStatus {
    code: 0;
    message: string;
}
interface FailureStatus {
    /* in failure case code will be anything other than 0 */
    code: number;
    reason: string;
}
type ProcessStatus = SuccessStatus | FailureStatus;

function isSuccess(response: ProcessStatus): response is SuccessStatus {
    return 'message' in response;
}

declare const a: ProcessStatus;

if (a.code === 0){
    // this doesn't work since typescript can't determine that failure case can't be 0
    console.log(a.message); 
}
if ('message' in a){
    // this works and is verifiable by the compiler
    console.log(a.message);
}
if (isSuccess(a)){
    console.log(a.message); // works as intended.
}
```

It seems like it would be preferable to have the narrowing occur safely, and that seems to be the standard expectation of TS developers who are unfamiliar with the current behavior.

**Comment by tadhgmister** [comment]
@Masstronaut then consider a case where message and reason are both optional. ([Playground Link][newexample]) I do think my original example is sufficient to get the point across but I respect your need for an example that has no alternative.


[newexample]: https://www.typescriptlang.org/play/?#code/FASwdgLgpgTgZgQwMZQAQGUCuSUGdfoQISa6oDewq1qSA9gCZQBcqADFTQLZT4IDmUAPytcEGOH7AAvqEixEKVADEEIADaYYUQsVIVO1APQAqE6nCpEGrWiQJcdxmgDuG9agBGaBGACeEAAWkqh0QbCoQb7sqCZGhrTOrGCYXN4wCdoOdGAiqGISYFKyEH4ADmgACjB0eAREJGQAvBjYdbqNqAA+Kmqa2h2kANzAwHCYYEgQIDkWBG28uAAU2rhlOY6s1bWLg7gAlKyr62COc604uw36lDSo2iQwYPe8J44AdPRMqE2-7COyYBGIyRQJoOB0dTqOhuIpWCZTGanUJgdR+VAMEBwOARSzhRJgTHTWa+BiJLg8SC4UbjSbE56BUnqKAAJVeGygAH0kGCkABrADCziWCC2NXa1wOtxoWNQIs+zh+fzY+2ld2MIKCIDIDDovDAAHIIKgXHQYHz8uAlKUKrgkBIysb7IbjUxoDAuOA0FFjdZ+nYHAGXV40Bx1TR6KdIVB3tD+EsAETHDlWGpcVBlcWLZgJgA0qAQ7x4fEE+yGqAS0lQUHUZzV6sjuGj71gNRgiczO3woO0LmrMDbiSYeYLCqY+YTbiCqGLuAELBHhayTbAZcrMjGCPpqEZhOZbLWHM5pCgABVylB+JgEDAGCKxV36nopQlZUttVhLvgRftVQk7sCJpmnyZAOBY8iElADDvP+EYbM2caJsmpzgmmGZZvgOb5oWs7zmWFZ3FWNZ1rBGoFrWdBAeaoFkOA0CQQwpEEk2zItgOZodhhZBBL2-aDl8UCLmOUATlOgQzos85YaOy45Ph66yLSiKzLuDD7uyKHcryfIAKIAB7atMRTKCANYMMsoqoNsErPn+dxvgauGCAa4EFnZ4aoIBWpkGIdBlGQjIAG4hPic48PkUBcL40xIBJvghLKjmSc55xgDCoSOkiCDqExjYIXQ8ZJhpZxwGhnZ1NJOHJVAa6EdWtZoPWAEgqSFhwKCaBOWg2oZfS2UmmgurBoIxq+OiYA3jUsL8DusBoLgVremCBaeI4kxoExgF0O1+JJSWUAuXApnqGSPVpcauCYNiIBIKZkCRJRE3sX2ECUfirZmrQgYwR5zHNh97YJuViw9lAfYA0OgnYcJokgNOXWVe8smriMhEbkAA
